crates/sui-framework/packages/sui-framework/sources/random.move

```diff
@@ -129,4 +131,146 @@ module sui::random {
     ) {
         update_randomness_state(self, new_round, new_bytes, ctx);
     }
+
+
+    /// Unique randomness generator, derived from the global randomness.
+    struct RandomGenerator has drop {
+        seed: vector<u8>,
+        counter: u16,
+        buffer: vector<u8>,
+    }
+
+    /// Create a generator. Can be used to derive up to MAX_U16 * 32 random bytes.
+    public fun new_generator(r: &Random, ctx: &mut TxContext): RandomGenerator {
+        let inner = load_inner(r);
+        let seed = hmac_sha3_256(&inner.random_bytes, &to_bytes(fresh_object_address(ctx)));
+        RandomGenerator { seed, counter: 0, buffer: vector[] }
+    }
+
+    // Get the next block of random bytes.
+    fun derive_next_block(g: &mut RandomGenerator): vector<u8> {
+        g.counter = g.counter + 1;
+        hmac_sha3_256(&g.seed, &bcs::to_bytes(&g.counter))
+    }
+
+    // Fill the generator's buffer with 32 random bytes.
+    fun fill_buffer(g: &mut RandomGenerator) {
+        let next_block = derive_next_block(g);
+        vector::append(&mut g.buffer, next_block);
+    }
+
+    /// Generate n random bytes.
+    public fun bytes(g: &mut RandomGenerator, num_of_bytes: u16): vector<u8> {
+        let result = vector[];
+        // Append RAND_OUTPUT_LEN size buffers directly without going through the generator's buffer.
+        let num_of_blocks = num_of_bytes / RAND_OUTPUT_LEN;
+        while (num_of_blocks > 0) {
+            vector::append(&mut result, derive_next_block(g));
+            num_of_blocks = num_of_blocks - 1;
+        };
+        // Take remaining bytes from the generator's buffer.
+        if (vector::length(&g.buffer) < ((num_of_bytes as u64) - vector::length(&result))) {
+            fill_buffer(g);
+        };
+        while (vector::length(&result) < (num_of_bytes as u64)) {
+            vector::push_back(&mut result, vector::pop_back(&mut g.buffer));
+        };
+        result
+    }
+
+    // Helper function that extracts the given number of bytes from the random generator and returns it as u256.
+    // Assumes that the caller has already checked that num_of_bytes is valid.
+    fun u256_from_bytes(g: &mut RandomGenerator, num_of_bytes: u8): u256 {
+        if (vector::length(&g.buffer) < (num_of_bytes as u64)) {
+            fill_buffer(g);
+        };
+        let result: u256 = 0;
+        let i = 0;
+        while (i < num_of_bytes) {
```

Comment:
@damirka / @manolisliolios, do we have a `vector<u8>` -> `u256 utility function that would help here (and can we add one if not)?

Reply:
I think I'd still need a loop here because I want to read exactly `num_of_bytes` bytes (something like `vector::pop_back_n(buffer, num_of_bytes)` would have helped for that)

Reply:
will update this code once we have a better alternative (hopefully next month)